upup/pkg/fi/nodeup/nodetasks/bootstrap_client.go

```diff
@@ -0,0 +1,193 @@
+	// CA is the CA certificate for kops-controller.
+	CA []byte
+	// Certs are the requested certificates.
+	Certs map[string]*BootstrapCert
+
+	client *http.Client
+	keys   map[string]*pki.PrivateKey
+}
+
+type BootstrapCert struct {
+	Cert *fi.TaskDependentResource
+	Key  *fi.TaskDependentResource
+}
+
+var _ fi.Task = &BootstrapClient{}
+var _ fi.HasName = &BootstrapClient{}
+var _ fi.HasDependencies = &BootstrapClient{}
+
+func (b *BootstrapClient) GetDependencies(tasks map[string]fi.Task) []fi.Task {
+	return nil
+}
+
+func (b *BootstrapClient) GetName() *string {
+	name := "BootstrapClient"
+	return &name
+}
+
+func (b *BootstrapClient) String() string {
+	return "BootstrapClient"
+}
+
+func (b *BootstrapClient) Run(c *fi.Context) error {
+	req := nodeup.BootstrapRequest{
+		APIVersion: nodeup.BootstrapAPIVersion,
+		Certs:      map[string]string{},
+	}
+
+	if b.keys == nil {
+		b.keys = map[string]*pki.PrivateKey{}
+	}
+
+	for name, certRequest := range b.Certs {
+		key, ok := b.keys[name]
+		if !ok {
+			var err error
+			key, err = pki.GeneratePrivateKey()
+			if err != nil {
+				return fmt.Errorf("generating private key: %v", err)
+			}
+
+			certRequest.Key.Resource = &asBytesResource{key}
+			b.keys[name] = key
+		}
+
+		pkData, err := x509.MarshalPKIXPublicKey(key.Key.(*rsa.PrivateKey).Public())
+		if err != nil {
+			return fmt.Errorf("marshalling public key: %v", err)
+		}
+		// TODO perhaps send a CSR instead to prove we own the private key?
+		req.Certs[name] = string(pem.EncodeToMemory(&pem.Block{Type: "RSA PUBLIC KEY", Bytes: pkData}))
+	}
+
+	resp, err := b.queryBootstrap(c, &req)
+	if err != nil {
+		return err
+	}
+
+	for name, certRequest := range b.Certs {
+		cert, ok := resp.Certs[name]
+		if !ok {
+			return fmt.Errorf("kops-controller did not return a %q certificate", name)
+		}
+		certificate, err := pki.ParsePEMCertificate([]byte(cert))
+		if err != nil {
+			return fmt.Errorf("parsing %q certificate: %v", name, err)
+		}
+		certRequest.Cert.Resource = asBytesResource{certificate}
+	}
+
+	return nil
+}
+
+func (b *BootstrapClient) queryBootstrap(c *fi.Context, req *nodeup.BootstrapRequest) (*nodeup.BootstrapResponse, error) {
+	if b.client == nil {
+		certPool := x509.NewCertPool()
+		certPool.AppendCertsFromPEM(b.CA)
+
+		b.client = &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					RootCAs:    certPool,
+					MinVersion: tls.VersionTLS12,
+				},
+			},
+		}
+	}
+
+	reqBytes, err := json.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+
+	bootstrapUrl := url.URL{
+		Scheme: "https",
+		Host:   net.JoinHostPort(c.Cluster.Spec.MasterInternalName, strconv.Itoa(wellknownports.KopsControllerPort)),
+		Path:   "/bootstrap",
+	}
+	httpReq, err := http.NewRequest("POST", bootstrapUrl.String(), bytes.NewReader(reqBytes))
+	if err != nil {
+		return nil, err
+	}
+	httpReq.Header.Set("Content-Type", "application/json")
+
+	token, err := b.Authenticator.CreateToken(reqBytes)
```

Comment:
Thanks for explaining the signature challenge.  I do understand, but ... can we implement this on GCE / OpenStack / Bare-Metal?  I know it's not _as_ simple, but could we do something like this:

```
type SignedBootstrapRequest struct {
  Request *BootstrapRequest `json:"request"` 
  AWSSecureTokenService *AWSSecureTokenService `json:"awsSecureTokenService"`
  // GCESignature *GCESignature
}

type AWSSecureTokenService struct {
  Headers map[string]string `json:"headers"`
}

```

It feels more extensible (for other clouds, possible for AWS also)

Reply:
I believe this interface is implementable by other clouds, assuming those clouds have some way to authenticate an instance. The interface I defined seems to me to have been reduced to the essentials: the provider-specific client gets the serialized body and produces an authenticator. The common code transfers the authenticator in the `Authorization:` header. The provider-specific server gets the serialized body and the authenticator and gets to say what the node name is or reject the request.

I'm presuming that multi-provider clusters are well out of scope.

Reply:
Agree that multi-provider clusters should be out of scope.  Even if we have a central management cluster, I think we would want node bring-up to be done relatively locally.  I could imagine forwarding the request to another service, but I imagine that would essentially be wrapping the current request anyway.

I can see the authorization header as isomorphic to what I described - we're encoding the signature data (with the body) into json, base64 encoding it, and then putting that in the authorization header.  I'm really not sure if that is simpler, but it's likely tied up in the skew problem ... I'll ponder!

Reply:
Putting the token in the header simplifies the task of getting the signature to include the request, since you can then just take a secure hash of the entire body. Putting it into the `Authorization:` header means anything in the middle is likely to treat it as confidential.

Reply:
I'm OK to proceed on this basis;  if we do find that on another cloud it's much trickier to put it into the header (e.g. GCE's TPM flow) then we might have to put it into the body for that cloud - is that reasonable?

Reply:
I think that's reasonable. If that were to happen, we'd probably do something like change the interfaces to return a modified body.